backend/provisioner/dev/dev_provisioner.go

```diff
@@ -0,0 +1,187 @@
+	err      error
+	done     atomic.Bool
+}
+
+// Provisioner is a provisioner for running FTL locally
+type Provisioner struct {
+	running     map[string]*task
+	postgresDSN string
+
+	postgresImage string
+	postgresPort  int
+}
+
+func NewProvisioner(postgresImage string, postgresPort int) *Provisioner {
+	return &Provisioner{
+		postgresImage: postgresImage,
+		postgresPort:  postgresPort,
+	}
+}
+
+var _ provisionerconnect.ProvisionerPluginServiceClient = (*Provisioner)(nil)
+
+func (d *Provisioner) Ping(context.Context, *connect.Request[ftlv1.PingRequest]) (*connect.Response[ftlv1.PingResponse], error) {
+	return &connect.Response[ftlv1.PingResponse]{}, nil
+}
+
+func (d *Provisioner) Plan(context.Context, *connect.Request[provisioner.PlanRequest]) (*connect.Response[provisioner.PlanResponse], error) {
+	panic("unimplemented")
+}
+
+func (d *Provisioner) Provision(ctx context.Context, req *connect.Request[provisioner.ProvisionRequest]) (*connect.Response[provisioner.ProvisionResponse], error) {
+	if d.running == nil {
+		d.running = map[string]*task{}
+	}
+
+	previous := map[string]*provisioner.Resource{}
+	for _, r := range req.Msg.ExistingResources {
+		previous[r.ResourceId] = r
+	}
+
+	task := &task{}
+	for _, r := range req.Msg.DesiredResources {
+		if _, ok := previous[r.Resource.ResourceId]; !ok {
+			switch tr := r.Resource.Resource.(type) {
+			case *provisioner.Resource_Postgres:
+				step := &step{resource: r.Resource}
+				task.steps = append(task.steps, step)
+
+				d.provisionPostgres(ctx, tr, req.Msg.Module, r.Resource.ResourceId, step)
+			default:
+			}
+		}
+	}
+
+	if len(task.steps) == 0 {
+		return connect.NewResponse(&provisioner.ProvisionResponse{
+			Status: provisioner.ProvisionResponse_NO_CHANGES,
+		}), nil
+	}
+
+	token := strconv.Itoa(rand.Int()) //nolint:gosec
+	d.running[token] = task
+
+	return connect.NewResponse(&provisioner.ProvisionResponse{
+		ProvisioningToken: token,
+		Status:            provisioner.ProvisionResponse_SUBMITTED,
+	}), nil
+}
+
+func (d *Provisioner) Status(ctx context.Context, req *connect.Request[provisioner.StatusRequest]) (*connect.Response[provisioner.StatusResponse], error) {
+	token := req.Msg.ProvisioningToken
+	task, ok := d.running[token]
+	if !ok {
+		return statusFailure("unknown token")
+	}
+	if !task.Done() {
+		return connect.NewResponse(&provisioner.StatusResponse{
+			Status: &provisioner.StatusResponse_Running{},
+		}), nil
+	}
+
+	var resources []*provisioner.Resource
+	for _, step := range task.steps {
+		if step.err != nil {
+			return statusFailure(step.err.Error())
+		}
+		resources = append(resources, step.resource)
+	}
+	delete(d.running, token)
+
+	return connect.NewResponse(&provisioner.StatusResponse{
+		Status: &provisioner.StatusResponse_Success{
+			Success: &provisioner.StatusResponse_ProvisioningSuccess{
+				UpdatedResources: resources,
+			},
+		},
+	}), nil
+}
+
+func statusFailure(message string) (*connect.Response[provisioner.StatusResponse], error) {
```

Comment:
Should we be using gRPC error responses here instead, or is there some reason to have our own?

Reply:
Ah, I think I went this way because the error is not directly related to the grpc call, but a result of the background process. But using grpc errors makes this cleaner. Will refactor 👍